Tetragrama/Components/LogUIComponent.cpp

```diff
@@ -26,38 +27,71 @@ namespace Tetragrama::Components
 
     void LogUIComponent::Render(ZEngine::Rendering::Renderers::GraphicRenderer* const renderer, ZEngine::Rendering::Buffers::CommandBuffer* const command_buffer)
     {
-        ImGui::Begin(m_name.c_str(), (m_can_be_closed ? &m_can_be_closed : NULL), ImGuiWindowFlags_NoCollapse);
+        ImGui::Begin(m_name.c_str(), (m_can_be_closed ? &m_can_be_closed : nullptr), ImGuiWindowFlags_NoCollapse);
 
-        ImGui::SameLine();
-        m_is_clear_button_pressed = ImGui::Button("Clear");
+        static const char* items[]      = {"All", "info", "error", "warn", "critical", "trace"};
+        static int         current_item = 0;
 
         ImGui::SameLine();
-        m_is_copy_button_pressed = ImGui::Button("Copy");
+        ImGui::SetNextItemWidth(70);
+        if (ImGui::BeginCombo("##Dropdown", items[current_item]))
+        {
+            for (int n = 0; n < IM_ARRAYSIZE(items); n++)
+            {
+                if (ImGui::Selectable(items[n], current_item == n))
+                {
+                    current_item = n;
+                    ImGui::SetItemDefaultFocus();
+                }
+            }
+            ImGui::EndCombo();
+        }
 
-        ImGui::Separator();
+        ImGui::SameLine();
+        if (ImGui::Button("Clear"))
+        {
+            ClearLog();
+            m_search_buffer[0] = '\0';
+        }
 
-        if (m_is_copy_button_pressed)
+        ImGui::SameLine();
+        if (m_is_copy_button_pressed = ImGui::Button("Copy"))
         {
             ImGui::LogToClipboard();
         }
 
-        if (m_is_clear_button_pressed)
+        ImGui::SameLine();
+        ImGui::InputTextWithHint("##Search", "Search logs...", m_search_buffer, IM_ARRAYSIZE(m_search_buffer));
+        ImGui::Separator();
+
+        std::string search_term;
+        if (std::strlen(m_search_buffer) > 0)
         {
-            ClearLog();
+            search_term = m_search_buffer;
+            std::transform(search_term.begin(), search_term.end(), search_term.begin(), ::tolower);
         }
 
         ImGui::PushStyleVar(ImGuiStyleVar_ItemSpacing, ImVec2(0, 0));
-
         if (ImGui::BeginTable("log_table", 1, ImGuiTableFlags_RowBg | ImGuiTableFlags_ScrollX | ImGuiTableFlags_ScrollY))
         {
             std::lock_guard<std::mutex> lock(m_mutex);
+            for (const auto& message : m_log_queue)
```

Comment:
can we switch m_log_queue type from vector to ThreadSafeQueue ? and pop available log message per frame?
this will reduce the pression on this component to render content